### PR TITLE
[autoscaler] Staroid node provider

### PR DESCRIPTION
**Pull request by Leemoonsoo**
## Why are these changes needed?

This pull request implements a node provider for [Staroid](https://staroid.com) cloud platform.
It enables to ray cluster be created by running

```bash
$ export STAROID_ACCESS_TOKEN=<...>
$ ray up python/ray/autoscaler/staroid/example-full.yaml
```

The provider code is based on Kubernetes node provider while Staroid is providing Kubernetes namespaces on the cloud, with some differences such as creating Kubernetes namespace using Staroid api, connecting Kubernetes API server remotely, and supporting non-root container.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by richardliaw** [comment]
@ericl feel free to remove assignment; just wanted to get your thoughts on where a third-party contrib for node-provider would go.

**Comment by ericl** [comment]
_private/staroid? I think it's fine to put in the same place as other cloud providers for now, there aren't that many that organization is an issue.

**Comment by Leemoonsoo** [comment]
Thanks for taking a look! Let me rebase and move the provider code to _private, and then ping to get a further review.

**Comment by Leemoonsoo** [comment]
@richardliaw Rebased and moved provider code to _private. I think it is ready for review. And please let me know if the code or doc location needs to be changed. :)

File: python/ray/autoscaler/_private/staroid/node_provider.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+import requests
+from staroid import Staroid
+from kubernetes import client, config
+import socket
+from contextlib import closing
+
+from uuid import uuid4
+from kubernetes.client.rest import ApiException
+
+from .command_runner import StaroidCommandRunner
```
Comment (richardliaw) [inline python/ray/autoscaler/_private/staroid/node_provider.py:13]: can you use absolute path instead here?

File: doc/source/cluster/cloud.rst [hunk 2]
```diff
@@ -145,6 +145,37 @@ Test that it works by running the following commands from your local machine:
     # Tear down the cluster.
     $ ray down ray/python/ray/autoscaler/gcp/example-full.yaml
 
+Staroid
+-------
+
+First, install the staroid client package (``pip install staroid``) then get `access token <https://staroid.com/settings/accesstokens>`_.
```
Comment (richardliaw) [inline doc/source/cluster/cloud.rst:151]: Can you move this under the Kubernetes section?

**Comment by richardliaw** [comment]
This looks great to me. Can you address the comments and ping me when done?

**Comment by Leemoonsoo** [comment]
Addressed your comments. Please check when you have a moment. Thank @richardliaw for review.

**Review by richardliaw — APPROVED** [review]

**Pull request merged**